File: BackEnd/src/main/java/com/platepilote/platepilote/budget/domain/repository/BudgetRepository.java
package com.platepilote.platepilote.budget.domain.repository;

import com.platepilote.platepilote.budget.domain.entity.Budget;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import java.util.List;
import java.util.UUID;

/**
 * Repository pour l'accès aux données des budgets.
 */
public interface BudgetRepository extends JpaRepository<Budget, UUID> {

    /**
     * Récupère les budgets actifs d'un utilisateur avec pagination.
     *
     * @param userId   identifiant de l'utilisateur
     * @param pageable paramètres de pagination
     * @return page des budgets
     */
    Page<Budget> findByUserIdAndDeletedAtIsNull(UUID userId, Pageable pageable);

    /**
     * Récupère tous les budgets actifs d'un utilisateur.
     *
     * @param userId identifiant de l'utilisateur
     * @return liste des budgets
     */
    List<Budget> findByUserIdAndDeletedAtIsNull(UUID userId);
}

File: FrontEnd/lib/features/budget/budget_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/network/api_client.dart';
import '../../core/network/api_response.dart';
import '../../core/repositories/base_repository.dart';
import '../../shared/models/budget.dart';

/// Repository des opérations liées au budget.
class BudgetRepository extends BaseRepository {
  BudgetRepository(super.apiClient);

  /// Liste les budgets avec pagination.
  Future<PageResponse<Budget>> listBudgets({
    int page = 0,
    int size = 20,
  }) async {
    try {
      final response = await apiClient.get(
        '/budgets',
        query: {'page': page, 'size': size},
      );
      return handlePageResponse(response, Budget.fromJson);
    } on DioException catch (e) {
      throw ApiException(extractMessage(e), e.response?.statusCode);
    }
  }

  /// Crée un nouveau budget.
  Future<Budget> createBudget({
    required double amount,
    String currency = 'USD',
    required String period,
    required String startDate,
    String? endDate,
  }) async {
    try {
      final response = await apiClient.post('/budgets', data: {
        'amount': amount,
        'currency': currency,
        'period': period,
        'startDate': startDate,
        'endDate': endDate,
      });
      return handleResponse(response, Budget.fromJson);
    } on DioException catch (e) {
      throw ApiException(extractMessage(e), e.response?.statusCode);
    }
  }

  /// Supprime un budget par son ID.
  Future<void> deleteBudget(String budgetId) async {
    try {
      await apiClient.delete('/budgets/$budgetId');
    } on DioException catch (e) {
      throw ApiException(extractMessage(e), e.response?.statusCode);
    }
  }

  /// Récupère les données d'analyse du budget.
  Future<Map<String, dynamic>> getAnalytics() async {
    try {
      final response = await apiClient.get('/budgets/analytics');
      final data = response.data;
      if (data is Map && data['data'] is Map) {
        return data['data'] as Map<String, dynamic>;
      }
      return {};
    } on DioException catch (e) {
      throw ApiException(extractMessage(e), e.response?.statusCode);
    }
  }

  /// Récupère les données d'économies.
  Future<Map<String, dynamic>> getSavings() async {
    try {
      final response = await apiClient.get('/budgets/savings');
      final data = response.data;
      if (data is Map && data['data'] is Map) {
        return data['data'] as Map<String, dynamic>;
      }
      return {};
    } on DioException catch (e) {
      throw ApiException(extractMessage(e), e.response?.statusCode);
    }
  }
}

/// Provider Riverpod pour [BudgetRepository].
final budgetRepositoryProvider = Provider<BudgetRepository>((ref) {
  return BudgetRepository(ref.watch(apiClientProvider));
});
